### PR TITLE
Support `features` key in `version` JSON output

### PR DESCRIPTION
**Pull request by mbg**
Add general support for extracting the `features` key in the output of the `version` command. 

### Merge / deployment checklist

- [ ] Confirm this change is backwards compatible with existing workflows.
- [ ] Confirm the [readme](https://github.com/github/codeql-action/blob/main/README.md) has been updated if necessary.
- [ ] Confirm the [changelog](https://github.com/github/codeql-action/blob/main/CHANGELOG.md) has been updated if necessary.

### CODE REVIEW

File: src/codeql.ts [hunk 1]
```diff
@@ -212,6 +212,19 @@ export interface CodeQL {
...
+  productName: string;
+  vendor: string;
+  version: string;
+  sha: string;
+  branches: string[];
+  copyright: string;
+  unpackedLocation: string;
+  configFileLocation: string;
+  configFileFound: boolean;
+  features?: { [name: string]: boolean };
```
Comment (henrymercer) [inline src/codeql.ts:225]: We can simplify this down to just the entries the CodeQL Action cares about:

```suggestion
  version: string;
  features?: { [name: string]: boolean };
```

File: src/codeql.ts [hunk 2]
```diff
@@ -430,6 +443,21 @@ function resolveFunction<T>(
...
+  vendor: "GitHub",
+  sha: "",
+  branches: [],
+  copyright: "",
+  unpackedLocation: "",
+  configFileLocation: "",
+  configFileFound: false,
+  version,
+});
+
```
Comment (henrymercer) [inline src/codeql.ts:460]: Minor: Move this to `testing-utils.ts`

File: src/codeql.ts [hunk 3]
```diff
@@ -212,6 +212,19 @@ export interface CodeQL {
   resolveExtractor(language: Language): Promise<string>;
 }
 
+export interface VersionOutput {
```
Comment (henrymercer) [inline src/codeql.ts:215]: Nit: rename to `VersionInfo`?

File: src/codeql.ts [hunk 4]
```diff
@@ -528,7 +556,9 @@ export async function getCodeQLForCmd(
     async getVersion() {
       let result = util.getCachedCodeQlVersion();
       if (result === undefined) {
-        result = (await runTool(cmd, ["version", "--format=terse"])).trim();
+        result = JSON.parse(
+          await runTool(cmd, ["version", "--format=json"]),
+        ) as VersionOutput;
```
Comment (henrymercer) [inline src/codeql.ts:561]: Probably worth wrapping this in a `try/catch` to handle unexpected output from `codeql version --format=json`.

**Review by henrymercer — COMMENTED** [review]
Nice!  A couple of minor comments.

File: src/codeql.ts (hunk 3 above)
Comment (mbg) [reply src/codeql.ts:215]: I went for `VersionOutput` to be consistent with the other interface names, but agree that `VersionInfo` is nicer. Consistency or nice name?

Comment (henrymercer) [reply src/codeql.ts:215]: Ah good point — I missed that while looking at just the diff.  I mildly prefer `VersionInfo`, but either is fine.

File: src/testing-utils.ts [hunk 5]
```diff
@@ -208,10 +208,17 @@ export function mockLanguagesInRepo(languages: string[]) {
   return listLanguages;
 }
 
-export function mockCodeQLVersion(version) {
+/**
+ * Constructs a `VersionInfo` object for testing purposes only.
+ */
+export const makeVersionInfo = (version: string): CodeQL.VersionInfo => ({
```
Comment (henrymercer) [inline src/testing-utils.ts:214]: Non-blocking: Out of curiosity, why convert to `export const`?

**Review by henrymercer — APPROVED** [review]

File: src/testing-utils.ts (hunk 5 above)
Comment (mbg) [reply src/testing-utils.ts:214]: (It was `const` from the start.)

Mostly a style preference coming from functional programming languages and I find the semantics end up causing me to shoot myself in the foot less often, but not sure there's much of a concrete benefit in this particular instance. Happy to change it to `function` if you have some good reasons (debugging?) for having that.

**Pull request merged**